File: datastructures/stack/stack-linked-list.py
class StackNode:
    def __init__(self, data: int = 0, next_node=None) -> None:
        self.data = data
        self.next: StackNode = next_node


class StackList:
    def __init__(self) -> None:
        self._top: StackNode = None
        self.size: int = 0

    def is_empty(self)->bool:
        return self.size == 0

    def push(self, x: int):
        if self.is_empty():
            self._top = StackNode(x)
            self.size += 1
            return

        newNode = StackNode(x)
        newNode.next = self._top
        self._top = newNode
        self.size += 1

    def pop(self):
        if self.is_empty():
            print("Stack is empty.\n")
            return
        
        self._top=self._top.next
        self.size-=1

    def top(self)->int:
        if self.is_empty():
            print("Stack is empty.\n")
            return -1
        
        return self._top.data

    def get_size(self)->int:
        return self.size
    


if __name__=="__main__":
    s=StackList()
    s.pop()
    s.push(32)
    s.push(4)
    print(f"\nTop: {s.top()}\nSize: {s.get_size()}")
    s.push(15)
    print(f"\nTop: {s.top()}\nSize: {s.get_size()}")
    s.pop()
    s.pop()
    print(f"\nTop: {s.top()}\nSize: {s.get_size()}")
    s.pop()
    print(f"\nTop: {s.top()}\nSize: {s.get_size()}")

    

File: datastructures/stack/stack-linked-list.cpp
#include <iostream>

class StackList;

class StackNode
{
private:
    int data;
    StackNode *next;

public:
    StackNode() : data(0)
    {
        next = 0;
    }
    StackNode(int x) : data(x)
    {
        next = 0;
    }
    StackNode(int x, StackNode *nextNode) : data(x), next(nextNode){};
    friend class StackList;
};

class StackList
{
private:
    StackNode *top;
    int size;

public:
    StackList() : size(0), top(0){};
    void Push(int x);
    void Pop();
    bool IsEmpty();
    int Top();
    int getSize();
};

void StackList::Push(int x)
{
    if (IsEmpty())
    {
        top = new StackNode(x);
        size++;
        return;
    }

    StackNode *newNode = new StackNode(x);
    newNode->next = top; // Push front
    top = newNode;
    size++;
}

void StackList::Pop()
{
    if (IsEmpty())
    {
        std::cout << "Stack is empty.\n";
        return;
    }

    StackNode *deleteNode = top;
    top = top->next;
    delete deleteNode;
    deleteNode = 0;
    size--;
}

bool StackList::IsEmpty()
{
    return (size == 0);
}

int StackList::Top()
{
    if (IsEmpty())
    {
        std::cout << "Stack is empty.\n";
        return -1;
    }
    return top->data;
}

int StackList::getSize()
{
    return size;
}

int main()
{
    StackList s;
    s.Pop();
    s.Push(32);
    s.Push(4);
    std::cout << "\ntop: " << s.Top() << "\nsize: " << s.getSize() << std::endl;
    s.Push(15);
    std::cout << "\ntop: " << s.Top() << "\nsize: " << s.getSize() << std::endl;
    s.Pop();
    s.Pop();
    std::cout << "\ntop: " << s.Top() << "\nsize: " << s.getSize() << std::endl;
    s.Pop();
    std::cout << "\ntop: " << s.Top() << "\nsize: " << s.getSize() << std::endl;

    return 0;
}
